fs.go

```diff
@@ -0,0 +1,183 @@
+type MutableProvider interface {
+	Provider
+
+	// BaseDir returns the base registry directory (when available).
+	BaseDir() string
+
+	// Init initializes a new registry in the local filesystem.
+	Init() error
+
+	// Verify verifies the integrity of the whole registry.
+	Verify() error
+
+	// UpdateEntity updates entity metadata in the registry.
+	UpdateEntity(entity *SignedEntityMetadata) error
+}
+
+type fsProvider struct {
+	baseDir string
+	fs      billy.Filesystem
+}
+
+func (p *fsProvider) getEntityPath(id signature.PublicKey) string {
+	entityID := publicKeyToFilename(id)
+	return p.fs.Join(registryDir, registryEntityDir, entityID+statementExt)
+}
+
+// Implements Provider.
+func (p *fsProvider) GetEntity(ctx context.Context, id signature.PublicKey) (*EntityMetadata, error) {
+	f, err := p.fs.Open(p.getEntityPath(id))
+	switch {
+	case err == nil:
+	case os.IsNotExist(err):
+		return nil, ErrNoSuchEntity
+	default:
+		return nil, fmt.Errorf("%w: failed to open entity metadata: %s", ErrCorruptedRegistry, err)
+	}
+	defer f.Close()
+
+	entity := new(EntityMetadata)
+	return entity, entity.Load(id, f)
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) BaseDir() string {
+	return p.baseDir
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) Init() error {
+	paths := []string{
+		registryDir,
+		p.fs.Join(registryDir, registryEntityDir),
+	}
+
+	for _, path := range paths {
+		if _, err := p.fs.Stat(path); !os.IsNotExist(err) {
+			return fmt.Errorf("registry already initialized (or corrupted)")
+		}
+
+		if err := p.fs.MkdirAll(path, 0o755); err != nil {
+			return fmt.Errorf("failed to create path %s: %w", path, err)
+		}
+
+		placeholderPath := p.fs.Join(path, placeholderFilename)
+		f, err := p.fs.OpenFile(placeholderPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o644)
+		if err != nil {
+			return fmt.Errorf("failed to create placeholder %s: %w", placeholderPath, err)
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) Verify() error {
+	// Verify entities.
+	entities, err := p.fs.ReadDir(p.fs.Join(registryDir, registryEntityDir))
+	if err != nil {
+		return fmt.Errorf("failed to read entity directory: %w", err)
+	}
+
+	for _, fi := range entities {
+		if filepath.Ext(fi.Name()) != statementExt {
+			continue
+		}
+
+		var id signature.PublicKey
+		if err = id.UnmarshalHex(strings.TrimSuffix(fi.Name(), statementExt)); err != nil {
+			return fmt.Errorf("entity: bad statement filename '%s': %w", fi.Name(), err)
+		}
+
+		if fi.Size() > MaxStatementSize {
+			return fmt.Errorf("entity: statement too big (size: %d max: %d): %s", fi.Size(), MaxStatementSize, fi.Name())
+		}
+
+		if _, err = p.GetEntity(context.Background(), id); err != nil {
+			return fmt.Errorf("entity: bad statement '%s': %w", fi.Name(), err)
+		}
+	}
+	return nil
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) UpdateEntity(entity *SignedEntityMetadata) error {
```

Comment:
(Commenting here since there's no PR to comment on in: https://github.com/oasisprotocol/metadata-registry)

I assume the flow with: https://github.com/oasisprotocol/metadata-registry will look something like:
1) fork repo
2) locally update your entity
3) submit PR

We should probably add a PR check that runs `Verify` and (probably) another one that checks that the `Serial` was updated (if the first check succeeds, and if we want to enforce this). Then most of the metadata submission can be automated. Maybe open issues (in registry repo) for this so we don't forget :-)

Reply:
Yeah that's indeed the idea of how this would be used.

Reply:
Added https://github.com/oasisprotocol/metadata-registry/issues/1.

Reply:
thanks!